### PR TITLE
Provide ability to use AtomGroup as selection for Contact

### LINKED ISSUE #2666: Provide AtomGroup selections to Contacts
## Feature request ##
_Originally posted by @orbeckst in https://github.com/MDAnalysis/mdanalysis/issues/2368#issuecomment-545079547_

* Add AtomGroup functionality:
  - It is  clunky that I have to provide *selections* instead of AtomGroups for `Contacts(..., selections=...))` to indicate which groups the contacts are formed between. I should be able to use static AtomGroups (cannot be dynamic). See #2369 

* Tidy up the helper function:
  - The https://github.com/MDAnalysis/mdanalysis/blob/9372d093acebb78bc5b1a2b4dba815a4e3281b04/package/MDAnalysis/analysis/contacts.py#L464 helper function can just use `Universe.copy()`.

### PR DESCRIPTION
**Pull request by shudipto-amin**
Fixes #2666

Summary
Solves the GSoC starter issue [#2666](https://github.com/MDAnalysis/mdanalysis/issues/2666)

Changes made in this Pull Request:
 - the class `mdanalysis/package/MDAnalysis/analysis/contacts.py::Contacts` is modified to accept static AtomGroups
 - the helper function `_new_selections()` in the same script is modified to use `Universe.copy()` **UPDATE: this has now been raised in new Issue #3570. This PR has reverted it back to the original.**
 - tests were added to `test_contacts.py` to ensure Contacts successfully uses static AtomGroups as well as selection texts.
 - tests were also added to ensure it fails when wrong types are provided for selection, including updating AtomGroups.

PR Checklist
------------
 - [x] Codecov coverage?
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: rather than using `xfail` it would be much better to use the `with pytest.raises()` construct with a match to the error message you are expecting (`select_error_message`).

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: same here with pytest.raises

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: same here with pytest.raises

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: Can i just query why the choice of `sel_acidic` and `sel_basic`? perhaps I'm missing some context here.

File: package/CHANGELOG
Comment (hmacdope) [inline package/CHANGELOG]: `no` vs `now` typo

**Review by hmacdope — CHANGES_REQUESTED** [review]
This looks like real progress! See my comments for changes :)

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Unfortunately this is not how the changelog should be updated.

1. Update your branch so you have the latest version of the changelog
2. Add yourself at the current line 23 (on the line starting with my username)
3. Put fixes as a bullet point under Fixes (currently at line 28).

Comment (shudipto-amin) [reply package/CHANGELOG]: Thanks for your instructions! I had some confusion about this part, which is clearer now. I updated the CHANGELOG accordingly, let me know if you have other corrections/suggestions.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (tylerjereddy) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: Perhaps a bit awkward to catch all/generic `Exception`s in a test and manually raise a `TypeError` inside the test. Perhaps following the guidance related to `pytest.raises()` will clear that up to test for the expected errors in the code proper, since manually raising an error in the testsuite isn't of much use to end users of course.

If some of the parameters fail and others succeed you can xfail on a case-by-case basis as well.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: Thanks for all your comments! Let me answer this while I work on implementing pystest.raises(). 
I initially thought of using `ag_acidic` and `ag_basic`, where `ag` is for AtomGroup, but then I realized it doesn't have to be an AtomGroup, and can be a selection text. I thought `sel` would be more general. However, now I realize that it can be easily confused with `self.sel_basic` and `self.sel_acidic`. Should I change the variable names to `ag_*` instead of `sel_*`?

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: This is indeed a better approach! I updated the testing code to use `pytest.raises()`, and `pytest.mark.parametrize` to combine all expected failure tests in one function. As for referring to the `select_error_message`, I updated `contacts.Contacts` class to define it as a static attribute, and referred to it in `test_contacts.py` using `with pytest.raises(TypeError, match=contacts.Contacts.select_error_message)`. Is that okay, or is there a better way?

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: yeah that makes sense.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: No need to duplicate here, can just use `contacts.Contacts.select_error_message` directly in the `pytest.raises expression`

**Comment by hmacdope** [comment]
@shudipto-amin looking good! Would you be able to resolve all the comments you have addressed so I can see what is still outstanding?

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: Thanks for you comment! I updated the code to use `@pytest.mark.parametrization` decorator, and I believe that should bring it in line with the style of `contacts.py`. I'm a bit uncomfortable with how I had to use the decorator, but maybe that's okay? Should we resolve?

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: Please make sure you also review the other PEP8 issues that pep8speaks has highlighted: https://github.com/MDAnalysis/mdanalysis/pull/3565#issuecomment-1072770416

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Entries should be newest first in the changelog

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: MDAnalysis tries to follow PEP8 where possible, as part of it, the recommendation is that line continuation should [use implicit line continuation where possible](https://peps.python.org/pep-0008/#:~:text=The%20preferred%20way%20of%20wrapping,a%20backslash%20for%20line%20continuation).

```suggestion
    select_error_message = ("selection must be either string or a "
                            "static AtomGroup")
```

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: There's a lot of repeated code here, let's see if we can condense it a little bit:

How about something like:

```python
@staticmethod
def _get_atomgroup(u, entry):
    if isinstance(entry, str):
        return u.select_atoms(entry)
    elif:
        if isinstance(entry, AtomGroup):
            if isinstance(entry, UpdatingAtomGroup):
                errmsg = "UpdatingAtomGroups are not supported"
                raise TypeError(errmsg)
            else:
                return entry
    else:
        raise TypeError(f"unrecognised selection type {entry}")

# ... and now back in __init__

self.grA, self.grB = [_get_atomgroup(var) for var in select]
```

This reduces the number of lines and also means you can directly test the staticmethod without having to go via class creation.

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: This technically changes the behaviour of q1q2, you could consider a case where `u` a trajectory which has been transferred in-memory and offset by a certain value (let's say you've shifted all the atomic positions by 10 A). By doing a copy on that object you now have a new trajectory that has the same shift in the atomic positions, whilst the old code would have led to a brand new universe that doesn't have this shift in atomic positions.

Whilst the copy behaviour is the one we are likely to want, introducing new changes that break behaviour without announcing it to users is a bad idea, because _someone_ might actually be relying on this behaviour (even though it's broken).

I realise @lilyminium pointed to fixing this in the original issue, but I don't agree that we should make this without warning users beforehand. I suspect further discussion will be needed here, so in the interest of saving time I would ask that we just not address this at this time and revisit it in the original issue before / fix it in a future PR.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: I see what you're doing here, but I think it actually makes things harder to read, I would suggestion creating a new fixture that's just for dealing with atomgroups.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: If going by my suggestion above, you can just directly do tests that run over the staticmethod instead here, it should be a lot simpler a parametrize.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Some comments, the way you've done it isn't wrong/bad, but it could be refined a little bit to make things easier long term.

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py]: I see your point. From what you say, it does sound like it would be best to make a separate issue/PR out of it, and revert it back to original here. So should I go ahead and create a new issue for it and reference it here (as well as in #2666), or leave it to core developers to raise it?

**Comment by shudipto-amin** [comment]
> @shudipto-amin looking good! Would you be able to resolve all the comments you have addressed so I can see what is still outstanding?

@hmacdope Thanks for your comments! I have resolved those conversations with the latest commits.

File: package/CHANGELOG
Comment (shudipto-amin) [reply package/CHANGELOG]: Done, I put it at the top.

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py]: Fixed in commits made since this comment.

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py]: Done! I added your suggested method in the Contacts class.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: Done, all tests are done on new static method `Contacts._get_atomgroup()`

**Comment by shudipto-amin** [comment]
@hmacdope @IAlibay @tylerjereddy 
Thank you all for your constructive comments! I learned a lot from them, and I believe I addressed all the concerns and the PR is ready or almost ready to merge. I just have a couple of questions:
1. According to [this discussion](https://github.com/MDAnalysis/mdanalysis/pull/3565#discussion_r831316421) with IAlibay, we should address the change to `universe.copy()` in a separate PR/Issue. Should I open a new issue for that?
2. The codecov report varies from time to time (without me pushing any new commits). Sometimes it shows the diff to be about -4%, sometimes 0%. Is there anything I can do to inverstigate/solve this, if it is a problem at all?

**Comment by IAlibay** [comment]
> @hmacdope @IAlibay @tylerjereddy Thank you all for your constructive comments! I learned a lot from them, and I believe I addressed all the concerns and the PR is ready or almost ready to merge. I just have a couple of questions

Thanks, one of the mentors will re-review over the next few days.

> 1. According to [this discussion](https://github.com/MDAnalysis/mdanalysis/pull/3565#discussion_r831316421) with IAlibay, we should address the change to `universe.copy()` in a separate PR/Issue. Should I open a new issue for that?

Please do, refer the original issue please.

> 2. The codecov report varies from time to time (without me pushing any new commits). Sometimes it shows the diff to be about -4%, sometimes 0%. Is there anything I can do to inverstigate/solve this, if it is a problem at all?

This is likely due to us needing to approve the majority of the gh action runners when you do a new commit. It _shouldn't_ be an issue but we will look into it as part of the review process.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (hmacdope) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py:188]: Could you add one more test, almost the same as this one but where you actually `.run()` the contacts with a `string` and `AtomGroup` input and compare the results?


You have tested the internal `staticmethod` `Contacts.get_atomgroup()` but not the actual flow of data through the whole function when a `string ` vs `AtomGroup` is used.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Almost there @shudipto-amin just one more test IMO then should be good to go. Thanks for your hard work.

**Comment by hmacdope** [comment]
Don't worry about codecov, it is just the runners not auto-triggering for first time contributors.

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: If possible could this be moved after `__init__`?

Usually folks inspecting classes will want to see the initialisation as the first thing so that they can quickly get an idea of what the class expects.

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py:391]: Please do add a `versionchanged` entry that mentions that AtomGroups are now supported.

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: This is now only needed within `_get_atomgroup` so I would just define it there rather than making it a class attribute.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: Improving readability is appreciated but I also don't think it's worth affecting the git history for an empty line.
```suggestion
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Couple more things

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py:391]: Done

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py:188]: Done! See [line 199:](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R199)
```python
def test_contacts_selections(self, universe):
```

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py]: I actually do use it in [this line](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R228) of `test_contacts.py`:
```python
       with pytest.raises(
            TypeError, match=contacts.Contacts._select_error_message
        ) as te:
            contacts.Contacts._get_atomgroup(universe, ag)
```
I can't think of another way to access that error message that doesn't require first calling a class/method. Is there some way, or do I just keep it as in internal attribute?

**Review by shudipto-amin — COMMENTED** [review]
@IAlibay I have addressed all but one comment below, regarding `_select_error_message` attribute, which I believe is still needed for testing functions.

@hmacdope I added the workflow test as you requested, [see line 199](https://github.com/MDAnalysis/mdanalysis/pull/3565/files#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R199):
```python
    def test_contacts_selections(self, universe):
```

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/contacts.py]: Honestly I think it's better to:
a) not have this as a class attribute
b) make the error messages within `_get_atomgroup` a bit more specific if you feel like it
c) just write a subsection of the message as part of the match in `test_select_wrong_types`

Convenience variables are great, but having them stick around for the single test just isn't worth it.

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: It's very pedantic but I'm going to object to these a little bit, this is not very readable.

If you're going to call run on it in one line then please re-arrange so that it's easier to read.

Otherwise close the brackets at 208 and call `cag.run()` separately.

Same thing below please.

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: Opted for rearrangement. Is this better?

File: package/MDAnalysis/analysis/contacts.py
Comment (shudipto-amin) [reply package/MDAnalysis/analysis/contacts.py]: Okay, done!

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: Sorry my readability issue wasn't with the way the arguments were arranged but instead with the `).run()` being alone on a single line. The compactness advantage of calling `run()` at the same time as class construction just disappears in this case.

Comment (shudipto-amin) [reply testsuite/MDAnalysisTests/analysis/test_contacts.py]: I think I get you now. I put `.run()` on a new line altogether in [this latest commit](https://github.com/MDAnalysis/mdanalysis/pull/3565/commits/55636bf304a590b0c6e52dfee76a13e0d4861a48#diff-c79b8e25558e1ba2f042d25eaa0e355502afba57cf03bf8c9b22e0c6103063a4R215).

File: package/MDAnalysis/analysis/contacts.py
Comment (IAlibay) [inline package/MDAnalysis/analysis/contacts.py]: Unfortunately read the docs hasn't triggered but there shouldn't be a blank space between :class: and the classname.
```suggestion
       :class:`Contacts` accepts both AtomGroup and string for `select`
```

File: package/CHANGELOG
Comment (IAlibay) [inline package/CHANGELOG]: Can you keep this empty line but remove the whitespace from it?

File: testsuite/MDAnalysisTests/analysis/test_contacts.py
Comment (IAlibay) [inline testsuite/MDAnalysisTests/analysis/test_contacts.py]: atomgroup comparison already calls np.array_equal in the background, doing a standard assert is orders of magnitude faster here (seconds to microseconds).
```suggestion
        assert ag_from_string == ag_from_ag
```

**Review by IAlibay — CHANGES_REQUESTED** [review]
Just a few tiny things and then I'm good with this.

@hmacdope do you want to re-review please?

File: package/CHANGELOG
Comment (shudipto-amin) [inline package/CHANGELOG]:
```suggestion
 
```

**Review by shudipto-amin — COMMENTED** [review]
Minor edits via suggestions.

**Review by hmacdope — APPROVED** [review]
Looks good to me @shudipto-amin. Thanks for all your hard work, we appreciate your contribution.

**Comment by shudipto-amin** [comment]
@IAlibay and @hmacdope, once again thanks for the feedback. Is there anything else, or is this ready to merge?

**Comment by IAlibay** [comment]
I think it's waiting on a review from me, will do in a little bit

**Comment by IAlibay** [comment]
Cycling PR to restart gh action runners that had halted last Friday (gh infrastructure issues it appears)

**Review by IAlibay — APPROVED** [review]
ltgm, thanks for the work on this @shudipto-amin.

**Pull request merged**
